src/core/thread/mesh_forwarder.cpp

```diff
@@ -192,7 +192,7 @@ void MeshForwarder::ClearChildIndirectMessages(Child &aChild)
 
         message->ClearChildMask(mNetif.GetMle().GetChildIndex(aChild));
 
-        if (!message->IsChildPending())
+        if (!message->IsChildPending() && !message->GetDirectTransmission())
```

Comment:
If we reached this part of the code (which is where we clear all (indirect) queued msg for a sleepy child if the child timed out and/or is removed from our child table).